pubsub/google/cloud/pubsub_v1/subscriber/_consumer.py

```diff
@@ -218,6 +223,54 @@ def _request_generator_thread(self):
             _LOGGER.debug('Sending request:\n%r', request)
             yield request
 
+    def _stop_request_generator(self, request_generator):
+        """Ensure a request generator is closed.
+
+        This **must** be done when recovering from a retry-able exception.
+        If not, then an inactive request generator (i.e. not attached to any
+        actual RPC) will be trying to access the same request queue as the
+        active request generator.
+
+        In addition, we want the gRPC thread consuming to cleanly exit so
+        that system resources are not wasted.
+
+        Args:
+            request_generator (Generator): A streaming pull request generator
+                returned from :meth:`_request_generator_thread`.
+
+        Returns:
+            bool: Indicates if the generator was successfully stopped. Will
+            be :data:`True` unless the queue is not empty and the generator
+            is running.
+        """
+        with self._put_lock:
+            try:
+                request_generator.close()
+            except ValueError:
+                # Should be ``ValueError('generator already executing')``
+                if not self._request_queue.empty():
+                    _LOGGER.debug(
+                        'Request generator could not be closed but '
+                        'request queue is not empty.')
+                    return False
+                # If we **cannot** close the request generator,
+                # then there is no blocking get on the queue. Since
+                # we have locked ``.put()`` this means that the
+                # queue **was** and remains empty.
+                self._request_queue.put(_helper_threads.STOP)
+                # Wait for the request generator to ``.get()`` the ``STOP``.
+                _LOGGER.debug(
+                    'Waiting for active request generator to receive STOP')
+                while not self._request_queue.empty():
+                    pass
+                request_generator.close()
```

Comment:
Luckily I re-factored the unit tests, because it made me realize that this could fail, because the generator still has to do:

```python
if request == _helper_threads.STOP:
    logger.debug()
    break
```

after the blocking `get()` completes, so this call to `close()` might happen **before** the generator gets those instructions.

I'm inclined to say we should either:

- Drop this statement completely (we are "safe" since we know that `STOP` has been pulled off the queue)
- Do a `while True` loop that tries to `close()` and `break`-s if it succeeds
- Do a `while request_generator.gi_running: pass` loop to block until it stops running and then call `close()`

Reply:
I am in favor of dropping `close()` as we know that as long as `STOP` (or *any* item in the queue) has been consumed post-RPC-cancellation that gRPC will exit the thread.

Reply:
Me too. Jinx.